### PR TITLE
New provider: search for shell commands explained

### PR DESCRIPTION
**Pull request by hackliff**
Hi there,

Quick but hopefully not so dirty PR to add [explainshell](explainshell.com).
I think I checked all the requirements of `CONTRIBUTING.md` but I'm not super familiar with OSS contribution so let me know ;)

I was planning to add [Regex101](https://regex101.com) and may be [Urban dictionnary](http://www.urbandictionary.com/) since it's so easy.

### CODE REVIEW

File: providers/explainshell/explainshell.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+	"fmt"
+	"strings"
+
+	"github.com/zquestz/s/providers"
+)
+
+// strings.Replace will apply for all matches when n < 0
+const REPLACE_ALL int = -1
+const PROVIDER_NAME = "explainshell"
+const PROVIDER_URL = "http://explainshell.com"
```
Comment (zquestz) [inline providers/explainshell/explainshell.go]: They offer https, therefore we should be using it. =)

File: providers/explainshell/explainshell.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/zquestz/s/providers"
+)
+
+// strings.Replace will apply for all matches when n < 0
+const REPLACE_ALL int = -1
```
Comment (zquestz) [inline providers/explainshell/explainshell.go]: Please use camelcase for constants and place them in a const block. Also only export them if they are required by other packages. Should look like:

```
const (
  replaceAll int = 1
  providerName = "explainshell"
  providerURL = "https://explainshell.com
)
```

File: providers/regex/regex.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package explainshell
+
+import (
+	"fmt"
+	"net/url"
+
+	"github.com/zquestz/s/providers"
+)
+
+const PROVIDER_NAME = "regex"
```
Comment (zquestz) [inline providers/regex/regex.go]: Please us the notation style above. This one was correctly using https. =)

File: providers/explainshell/explainshell.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+type Provider struct{}
+
+// BuildURI generates a search URL for Explain Shell
+func (p *Provider) BuildURI(q string) string {
+	searchParam := strings.Replace(q, " ", "+", REPLACE_ALL)
+	return fmt.Sprintf("%s/explain?cmd=%s", PROVIDER_URL, searchParam)
+}
+
+func (p *Provider) Tags() []string {
+	return []string{"shell", "tech"}
```
Comment (zquestz) [inline providers/explainshell/explainshell.go]: Should be in "code" tag. Try to keep to existing tags unless there is a good argument to be made for a new one.

File: providers/regex/regex.go [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+// Provider merely implements the Provider interface
+type Provider struct{}
+
+// BuildURI generates a search URL for Explain Shell
+func (p *Provider) BuildURI(q string) string {
+	return fmt.Sprintf("%s/library?orderBy=RELEVANCE&search=%s", PROVIDER_URL, url.QueryEscape(q))
+}
+
+func (p *Provider) Tags() []string {
+	return []string{"regex", "tech", "tool", "snippet"}
```
Comment (zquestz) [inline providers/regex/regex.go]: Should be in "code" tag just like explainshell.

File: Makefile [hunk 6]
```diff
@@ -1,8 +1,10 @@
 APPNAME = s
-OSARCH = darwin/386 darwin/amd64 dragonfly/amd64 freebsd/386 freebsd/amd64 freebsd/arm linux/386 linux/amd64 linux/arm netbsd/386 netbsd/amd64 netbsd/arm openbsd/386 openbsd/amd64 plan9/386 plan9/amd64 solaris/amd64 windows/386 windows/amd64
-DIRS = darwin_386 darwin_amd64 dragonfly_amd64 freebsd_386 freebsd_amd64 freebsd_arm linux_386 linux_amd64 linux_arm netbsd_386 netbsd_amd64 netbsd_arm openbsd_386 openbsd_amd64 plan9_386 plan9_amd64 solaris_amd64 windows_386 windows_amd64
 OUTDIR = pkg
 
+# allow to reduce cross compilation scope if needed
```
Comment (zquestz) [inline Makefile]: SUPERNIT: Allow user to override cross compilation scope.

**Comment by zquestz** [comment]
Thanks for the submission! These sites look like a good fit, just a few minor changes required before merge.

File: Makefile (hunk 6 above)
Comment (hackliff) [inline Makefile]: Not sure about what you mean here... do you want me to update the comment in the Makefile ?

**Comment by hackliff** [comment]
Hi @zquestz thanks for the review !

Except for the comment above I updated the code so it's ready to push as soon as I have your answer for the Makefile.

File: Makefile (hunk 6 above)
Comment (zquestz) [inline Makefile]: Yep that was why it was a super nit. =)

**Comment by zquestz** [comment]
This looks great, if you don't mind squashing the commits that would be fantastic. Then I will happily merge. =)

File: providers/regex/regex.go [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+)
+
+func init() {
+	providers.AddProvider(providerName, &Provider{})
+}
+
+// Provider merely implements the Provider interface
+type Provider struct{}
+
+// BuildURI generates a search URL for Explain Shell
```
Comment (zquestz) [inline providers/regex/regex.go]: NIT: This is Regex 101 not Explain Shell. =)

**Comment by hackliff** [comment]
Sorry about the micro-mistakes... Does the git history looks good ?

**Comment by zquestz** [comment]
Perfecto @hackliff. Thank you for taking the time to submit a PR and work with me on modifications!

**Pull request merged**